environment/src/main/java/com/sequenceiq/environment/environment/service/PollerCollection.java

```diff
@@ -0,0 +1,218 @@
+                    .collect(Collectors.toList());
+            return evaluateResult(pollingCrn, remaining, results);
+        };
+    }
+
+    private AttemptResult<Void> fetchStartDistroXResult(List<String> remaining, String crn) {
+        StackV4Response stack = distroXV1Endpoint.getByCrn(crn, Collections.emptySet());
+        if (stackAndClusterAvailable(stack, stack.getCluster())) {
+            return AttemptResults.finishWith(null);
+        } else {
+            remaining.add(crn);
+            return checkStartStatus(stack);
+        }
+    }
+
+    public AttemptMaker<Void> stopDatalakePoller(ArrayList<String> pollingCrn, Long envId) {
+        return () -> {
+            if (PollGroup.CANCELLED.equals(EnvironmentInMemoryStateStore.get(envId))) {
+                LOGGER.info("Datalake polling cancelled in inmemory store, id: " + envId);
+                return AttemptResults.breakFor("Datalake polling cancelled in inmemory store, id: " + envId);
+            }
+            List<String> remaining = new ArrayList<>();
+            List<AttemptResult<Void>> results = pollingCrn.stream()
+                    .map(crn -> fetchStopDatalakeResult(remaining, crn))
+                    .collect(Collectors.toList());
+            return evaluateResult(pollingCrn, remaining, results);
+        };
+    }
+
+    private AttemptResult<Void> fetchStopDatalakeResult(List<String> remaining, String crn) {
+        SdxClusterResponse sdx = sdxEndpoint.getByCrn(crn);
+        if (sdxStopped(sdx)) {
+            return AttemptResults.finishWith(null);
+        } else {
+            remaining.add(crn);
+            return checkStopStatus(sdx);
+        }
+    }
+
+    public AttemptMaker<Void> startDatalakePoller(ArrayList<String> pollingCrn, Long envId) {
+        return () -> {
+            if (PollGroup.CANCELLED.equals(EnvironmentInMemoryStateStore.get(envId))) {
+                LOGGER.info("Datalake polling cancelled in inmemory store, id: " + envId);
+                return AttemptResults.breakFor("Datalake polling cancelled in inmemory store, id: " + envId);
+            }
+            List<String> remaining = new ArrayList<>();
+            List<AttemptResult<Void>> results = pollingCrn.stream()
+                    .map(crn -> fetchStartDatalakeResult(crn, remaining))
+                    .collect(Collectors.toList());
+            return evaluateResult(pollingCrn, remaining, results);
+        };
+    }
+
+    private AttemptResult<Void> fetchStartDatalakeResult(String crn, List<String> remaining) {
+        SdxClusterResponse sdx = sdxEndpoint.getByCrn(crn);
+        if (sdxStarted(sdx)) {
+            return AttemptResults.finishWith(null);
+        } else {
+            remaining.add(crn);
+            return checkStartStatus(sdx);
+        }
+    }
+
+    @NotNull
+    private AttemptResult<Void> evaluateResult(ArrayList<String> pollingCrn, List<String> remaining, List<AttemptResult<Void>> results) {
```

Comment:
Why not using the Set interface instead of ArrayList? A set has the uniqueness attribute you need, and you don't need ordering what a list provides. I think the logic here should be rewritten to use the proper collection type (set).

Reply:
it is not an internal method but I don't want to use an unnecessary wrap here

Reply:
I cannot approve the PR. Code quality is more important than avoiding a wrapping.